LIPS/lip-composite_beacon_report_receiver.md

```diff
@@ -0,0 +1,158 @@
+
+Currently, the `LidoOracle` [contract](https://docs.lido.fi/contracts/lido-oracle) provides only one slot for a quorum report intercepting callback (see [LidoOracle.setBeaconReportReceiver](https://docs.lido.fi/contracts/lido-oracle#setbeaconreportreceiver)). We plan to occupy this slot soon with a newly developed contract implementing the [LIP-6 coverage application mechanism](https://research.lido.fi/t/lip-6-in-protocol-coverage-proposal/1468). In contrast, we have a technical vision of the future cross-chain/L2 [upgrades](https://hackmd.io/f3416OoaS2e1l2xu_bX6SA), which may require append additional callbacks propagating oracle beacon reports.
+
+## Specification
+
+We propose the following contracts interface. The code below presumes the Solidity v0.8 syntax.
+
+### `OrderedCallbacksArray` contract
+
+The contract provides callbacks storage in the form of addresses dynamic array.
+
+```solidity
+modifier onlyVoting()
+```
+Enforces the `msg.sender` to equal `voting` address set upon the contract construction.
+
+```solidity
+constructor(address _voting)
+```
+Init contract and store the provided `_voting` address to arm the `onlyVoting` modifier.
+See: `onlyVoting`.
+
+```solidity
+function addCallback(address _callback) external override onlyVoting;
+```
+Adds the provided `_callback` element at the end of the callbacks array, after its current last element.
+* Reverts if `_callback` address is zero.
+* Reverts if `msg.sender` is not equal to the stored `voting` address.
+* Reverts if the new length of the callbacks array becomes greater than the `MAX_CALLBACKS_CNT` contract-wide constant.
+* Emits the `CallbackAdded` event.
+See: `callbacksLength`.
+
+```solidity
+function insertCallback(address _callback, uint256 _atIndex) external override onlyVoting;
+```
+Inserts the `_callback` element at the specified by the `_atIndex` param location in the callbacks array.
+Elements at and after the `_atIndex` position are shifted one position right to preserve an existing invocation order.
+* Reverts if `_callback` address is zero.
+* Reverts if `msg.sender` is not equal to the stored `voting` address.
+* Reverts if `_atIndex` greater than the length of the callbacks array.
+* Reverts if new length of the callbacks array becomes greater than the `MAX_CALLBACKS_CNT` contract-wide constant.
+* Emits the `CallbackAdded` event.
+See: `callbacksLength`.
+
+```solidity
+function removeCallback(uint256 _atIndex) external override onlyVoting
+```
+Removes element with a given by the `_atIndex` param position from the callbacks array.
+Elements after the `_atIndex` position are shifted one position left to preserve an existing invocation order.
+* Reverts if `_atIndex` is equal to or greater than the length of the callbacks array.
+* Reverts if `msg.sender` is not equal to the stored `voting` address.
+* Emits the `CallbackRemoved` event.
+See: `callbacksLength`.
+
+```solidity
+function callbacks(uint256 _atIndex) external view returns (address);
```

Comment:
Imho, `getCallback` or `callbackAt` would be a clearer name

Reply:
Yeah, you are completely right.
But this one is autogenerated getter for the public `callbacks` storage var.
So I just decided to make it consistent from this side.